src/DataProtection/EntityFrameworkCore/src/EntityFrameworkCoreXmlRepository.cs

```diff
@@ -38,15 +38,15 @@ public EntityFrameworkCoreXmlRepository(IServiceProvider services, ILoggerFactor
         }
 
         /// <inheritdoc />
-        public virtual IReadOnlyCollection<XElement> GetAllElements()
+        public virtual IReadOnlyCollection<XElement?> GetAllElements()
         {
             using (var scope = _services.CreateScope())
             {
                 var context = scope.ServiceProvider.GetRequiredService<TContext>();
 
                 // Put logger in a local such that `this` isn't captured.
                 var logger = _logger;
-                return context.DataProtectionKeys.AsNoTracking().Select(key => TryParseKeyXml(key.Xml, logger)).ToList().AsReadOnly();
+                return context.DataProtectionKeys.AsNoTracking().Select(key => TryParseKeyXml(key.Xml!, logger)).ToList().AsReadOnly();
```

Comment:
Perhaps the way to do this would be to declare `DataProtectionKey.Xml` as non-null. While you could construct it with a null value, the framework doesn't. It helps reflect the intent rather than the implementation. What say?

Reply:
TryParseKeyXml could still return false if there is invalid XML.

I don't know enough about this API to say. SME should say whether it is ok to filter out non-valid keys.

Reply:
The expectation is that the return value of `GetAllElements` is an enumeration of non-null elements. This can also be seen via the call site within `XmlKeyManager`, which merrily dereferences properties on the returned elements with no null checks.

https://github.com/dotnet/aspnetcore/blob/6afd6f40c043f96c8fb1ee06cbebbc1d229efddc/src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs#L152-L161

Furthermore, the original design assumed that failures to parse the XML keyring were fatal and that corrupt XML files wouldn't be ignored. But this wasn't ever a steadfast rule. So presumably individual repositories could change the behavior if they have good reason to do so.

Reply:
👍 . I'll leave this API as non-nullable and summarize your response as a comment.

Reply:
I went ahead and updated the implementation to match the RegistryXmlRepo. It seems odd to swallow parse exceptions without good reason.